### PR TITLE
Base stamping decisions on --stamp configuration, not attributes on t…

### LINKED ISSUE #1451: stamping results in different digest with remote cache
Stamping a container image results in a different sha256 on each build. That may not be surprising in a local build scenario, but I expect the remote cache to serve up the files necessary to NOT generate a different sha256.

Repro steps:
```
bazel build --remote_http_cache //:image
cat bazel-bin/push_image.digest
bazel clean
bazel build --remote_http_cache //:image
cat bazel-bin/push_image.digest
```
Note that the digest is different between builds. I would expect the remote cache results in the same build stamped as before.

BUILD file:
```
container_image(
    name = "image",
    base = ":app_layer",
    stamp = True,
)

container_push(
    name = "push_image",
    format = "Docker",
    image = ":image",
    registry = "{STABLE_REGISTRY}",
    repository = "{STABLE_REPOSITORY_PATH}",
    tag = "{STABLE_IMAGE_TAG}",
)
```

### PR DESCRIPTION
**Pull request by alexeagle**
…he rule

This follows the example of rules_go and rules_nodejs.

It means we don't produce cache-busting outputs unless the user
explicitly requests non-deteriminism in the build with --stamp.

BREAKING CHANGE:
- To get stamped outputs, users must now call bazel with `--stamp`
- the stamp attribute was removed from some providers
- the stamp attribute was deprecated on container_push and container_bundle, it is now removed
- the stamp attribute was removed from container_image, there is no
replacement

Fixes #1451

### CODE REVIEW

**Review by gravypod — CHANGES_REQUESTED** [review]
Could we add an update to [README.md](BUILD_CONTEXT_ATTRS) that explains this behavior?

**Review by gravypod — APPROVED** [review]
I think this is good but as we discussed we should wait for a major version change whenever we're ready.

**Comment by alexeagle** [comment]
Okay so what's our cadence for major releases?
Since we haven't released a 1.0 semver dictates that breaking changes are in every minor. So we don't have a way to signal this beyond either just a 0.18.0 including this change.
I think we should have a 1.0. Could discuss doing that now? holding this change for that doesn't make a difference for semver though

**Comment by gravypod** [comment]
Sounds good. I can luck off a 1.0 on Monday if you're ok with that? Let's
also check in with pcj but in my opinion this is fine.

**Comment by alexeagle** [comment]
If we do 1.0 we need to say what that means and how we'll version in the future. Can discuss at our meeting Monday but I'll have to duck out early due to conflict

File: README.md [hunk 1]
```diff
@@ -250,14 +250,18 @@ to use `container_push` with custom docker authentication credentials.
...
+`container_push`, `container_bundle`, or `container_image` is to
 be able to vary the tag that is pushed or embedded.  There are two options
 at present for doing this.
 
 ### Stamping
 
-The first option is to use stamping. Stamping is enabled when a supported
-attribute contains a python format placeholder (e.g. `{BUILD_USER}`).
+The first option is to use stamping.
+Stamping is enabled when bazel is run with `--stamp`.
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply README.md:264]: Updated this to still have a `stamp` attribute on our three stamp-aware rules, and that attribute is now a trinary. So you can still use stamp=always behavior like before.

**Comment by alexeagle** [comment]
fyi @aptenodytes-forsteri rebased and landable.

**Pull request merged**